components/mdi/mdi_can_interface/Core/Src/can.c

```diff
@@ -0,0 +1,126 @@
+
+    output = ((float)pedal_data / (float) UINT32_MAX ) * (float)UINT10_MAX;
+	
+    if (output >  MAX_OUT) //cap the output voltage to our voltage limit
+        output =  MAX_OUT;
+    
+    return output; 
+}
+
+/**
+ * @brief Sends a desired voltage to the DAC, by transmitting the correct I2C message.
+ * @param uint16_t parsed_voltage should be a number between 0 and 1023. This value will be scaled to a corresponding voltage
+ * by the DAC, where 0 is 0V and 1023 is Vmax. This value is a relative porcentage value that comes from the parsed_voltage function. 
+ * @param DAC_ADDR: this is the I2C bit address of the DAC we want to send the I2C signal to
+ * @param *hi2c1 is the HAL handle for I2C
+ * @retval None
+ */
+void Send_Voltage(uint16_t parsed_voltage, uint8_t DAC_ADDR, I2C_HandleTypeDef *hi2c1)
+    {
+		uint8_t DAC_msg_buffer[2];
+		uint16_t dac_data;
+		if(parsed_voltage > 1023) parsed_voltage = 1023;
+    	dac_data = parsed_voltage << 2;
+    	DAC_msg_buffer[0] = dac_data >> 8;
+    	DAC_msg_buffer[1] = dac_data;
+    	HAL_I2C_Master_Transmit(hi2c1, DAC_ADDR, DAC_msg_buffer, BUFFER_SIZE, HAL_MAX_DELAY);
+    }
+
+/**
+ * @brief Takes the 0-1.0 number that represents the desired regen amount as a porcentage and sends it as a 10bit number to a DAC.
+ * @param regen: is float between 0 and 1.0 represeting a porcentage  
+ * @param DAC_REGEN_ADDR: this is the I2C bit address of the DAC we want to send the I2C signal to (in this case the DAC for the regen signal)
+ * @param *hi2c1 is the handle for the I2C
+ * @retval None
+ */
+void Send_Regen(float regen, uint8_t DAC_REGEN_ADDR, I2C_HandleTypeDef *hi2c1){
+
+	uint8_t DAC_msg_buffer[2];
+	uint16_t dac_data;
+	
+	if(regen > 1.0) regen = 1.0;
+
+	dac_data = regen * (float)UINT10_MAX; //converts to 10bit 
+	
+    	
+	dac_data =  dac_data << 2;
+    DAC_msg_buffer[0] = dac_data >> 8;
+    DAC_msg_buffer[1] = dac_data; 
+    HAL_I2C_Master_Transmit(hi2c1, DAC_REGEN_ADDR, DAC_msg_buffer, BUFFER_SIZE, HAL_MAX_DELAY);
+}
+/**
+ * @brief Takes the CAN message and decodes the data to know in which mode of operation the motor controller should be in and seperates the message data into
+ *  into the acceleration and velocity components 
+ * @param RxData: array of 8 bit wide elements that containes the CAN message sent by the SCN. 
+ * @param *CAN_msg: struct that holds the velocity, acceleration and current state of operation of the motor controller 
+ * @retval Modifies CAN_msg struct with the appropriate modes of operation and the values for velocity and accelation
+ */
+ void decode_CAN_velocity_message(uint8_t RxData[], CAN_msg_t* CAN_msg){
+
+	CAN_msg->velocity =  ( (int32_t) RxData[3] << 3*8) | ( (int32_t) RxData[2] << 2*8) | ( (int32_t) RxData[1] << 1*8) | ( (int32_t) RxData[0] << 0); 
+
+	CAN_msg->acceleration = (uint32_t) ( (uint32_t) RxData[7] << 3*8) | ( (uint32_t) RxData[6] << 2*8) | ( (uint32_t) RxData[5] << 1*8) | ((uint32_t) RxData[4] << 0);
```

Comment:
Nit: split these assignments over multiple lines for readability. Also consider wrapping each cast in parentheses to make it clear that it happens before the shift (although it's not necessary here, it's good for code clarity)

Also, not a problem here but FYI beware bitshifting signed types (like `int32_t`). If you do a right shift on a signed value, most C compilers will perform an arithmetic shift instead of a logical shift. If you're not expecting this, you have a bug on your hands! See https://en.wikipedia.org/wiki/Bitwise_operations_in_C#Right_shift_%3E%3E

```suggestion
	CAN_msg->velocity =
	    ( ((int32_t) RxData[3]) << 3*8) |
	    ( ((int32_t) RxData[2]) << 2*8) |
	    ( ((int32_t) RxData[1]) << 1*8) |
	    ( ((int32_t) RxData[0]) << 0); 

	CAN_msg->acceleration =
	    ( ((uint32_t) RxData[7]) << 3*8) |
	    ( ((uint32_t) RxData[6]) << 2*8) |
	    ( ((uint32_t) RxData[5]) << 1*8) |
	    ( ((uint32_t) RxData[4]) << 0);
```